### DISCUSSION TITLE
[v1] How best to deal with custom non-native inputs and non-string data?

### DISCUSSION THREAD

**Discussion by halljus** [edmundhung/conform#420] [Q&A]
We work with various custom components that represent non-native inputs and trying to find the best pattern for integrating with Conform v1. One example is https://github.com/ericgio/react-bootstrap-typeahead

Ultimately (in this example), the goal is to allow a user to select multiple items (e.g., pizza toppings) from a combobox and include an array of the selected item ids in the submission data. Without Zod and Conform, one might include these selected ids in the form data by rendering each one as an input with the corresponding id as the value but all sharing the same name. Then, you could pull them out with `formData.getAll('toppingId')`. We haven't figured out how to do this with Zod + Conform (the field list/array helpers don't seem applicable here, but maybe we're missing something).

In Conform v1, it seems a field value cannot be an array (at least not according to TypeScript and `defaultValue`), so we're calling `JSON.stringify` on the array of ids and storing that in a field value. But this means we also have to call `JSON.parse` to read it back out. This works, but we're just wondering if there's a better way?

FWIW, our schema looks like this:

```ts
// Goal: when it hits the action, `toppingIds` is an array of ids (strings) with length > 0
const schema = z
  .object({
    toppingIds: z.preprocess((value) => {
      if (!value) return [];
      return JSON.parse(String(value));
    }, z.string().array()),
  })
  .superRefine(({ toppingIds }, ctx) => {
    if (toppingIds.length === 0) {
      ctx.addIssue({
        path: ['toppingIds'],
        message: 'At least one topping is required',
        code: z.ZodIssueCode.custom,
      });
    }
  });
```

And our Conform initialization:

```ts
// defaultSelectedToppingIds => ['abc', 'def']

const [form, fields] = useForm({
  constraint: getZodConstraint(schema),
  defaultValue: {
    toppingIds: JSON.stringify(defaultSelectedToppingIds),
  },
  // ...
});

const toppingIdsControl = useInputControl({
  formId: form.id,
  name: fields.toppingIds.name,
  // Is this even needed? How does it relate to defaultValue above?
  initialValue: JSON.stringify(defaultSelectedToppingIds),
});
```

Then to read/write:

```tsx
<Typeahead
  selected={
    fields.toppingIds.value
      ? JSON.parse(fields.toppingIds.value).map((toppingId) =>
          toppingOptions.find((option) => option.value === toppingId),
        )
      : []
  }
  onChange={(selectedOptions) => {
    toppingIdsControl.change(JSON.stringify(selectedOptions.map((option) => option.value));
  }
  // ...
/>
```

**Comment by edmundhung** [comment]
You can model this with a multi select element like [this one](https://github.com/edmundhung/conform/discussions/389). Typescript will warn you if you do this now but it will work fine if you render the select element yourself. I also have a PR https://github.com/edmundhung/conform/pull/419 ready that should fix the issue and works without rendering a select manually.
